ranndy/models/vae.py

```diff
@@ -0,0 +1,76 @@
+# Based on: https://arxiv.org/pdf/1511.06349.pdf
+# TODO: Word dropout and historyless decoding
+
+from models.auto_encoder import SentenceAutoEncoder
+import tensorflow as tf
+
+
+class SentenceVAE(SentenceAutoEncoder):
+
+    def __init__(self, data_iterator, mode=tf.estimator.ModeKeys.TRAIN):
+
+        # Instance variables associated with VAE
+        self.encoding_mean = None
+        self.encoding_log_var = None # Note: Log(variance) is learned to ensure positivity
+        self.sampled_state = None
+        self.reconstruction_loss = None
+        self.kl_loss = None
+        self.kl_weight_scaling = 10 # TODO: Tune these values
+        self.kl_weight_offset = 5000
+
+        super().__init__(data_iterator, mode)
+
+    def _build_encoder(self):
+        self.encoder_cell = tf.nn.rnn_cell.LSTMCell(self.lstm_size, name="vae_encoder_cell")
+        output, encoder_state_tuple = tf.nn.dynamic_rnn(self.encoder_cell, self.enc_embedding_output,
+                                             sequence_length=self.iterator.source_length, dtype=tf.float32)
+
+        # Since encoder state is an LSTMStateTuple, we extract only the state
+        encoder_state = encoder_state_tuple.c
+
+        # Layers to convert output to mean and variance
+        #   Note: Paper uses linear layer (i.e. no activation)
+        self.encoding_mean = tf.layers.dense(encoder_state, self.lstm_size, activation=None, name="linear_mean_layer")
+        self.encoding_log_var = tf.layers.dense(encoder_state, self.lstm_size, activation=None, name="linear_var_layer")
+
```

Comment:
Why are the mean and variance the same? Don't we have to increase the dimenstionality of the encoder state to output both a vector of means and a vector of variances (log variances) instead of just one vector?

Reply:
They're not the same, they come from 2 different linear layers and thus produce 2 different vectors. The output is a vector based on self.lstm_size (this is the required size since it must match the decoder state size).